geps/gep-1767/index.md

```diff
@@ -0,0 +1,701 @@
+
+The server response for the CORS "preflight" request includes the following headers:
+    `Access-Control-Allow-Origin` response header indicates whether the response can be shared with requested resource from the given `Origin`.
+    `Access-Control-Allow-Methods` response header specifies one or more HTTP methods are accepted by the server when accessing the requested resource. 
+    `Access-Control-Allow-Headers` response header indicates which HTTP headers can be used during the actual cross-origin request.
+    
+The `Access-Control-Max-Age` optional response header indicates how long (in seconds) the information provided by the headers `Access-Control-Allow-Methods` and `Access-Control-Allow-Headers` can be cached by client. The default value for `Access-Control-Max-Age` is 5 seconds. Until the time specified by `Access-Control-Max-Age` elapses, the client doesn’t have to send another "preflight" request.
+
+The optional response header `Access-Control-Expose-Headers` controls which HTTP response headers are exposed to clients for the actual cross-origin request. 
+
+If the server specifies the response header `Access-Control-Allow-Credentials: true`, the actual cross-origin request will be able to use credentials for getting sensitive resources. 
+Credentials are cookies, TLS client certificates, or authentication headers containing a username and password.
+
+After the server has permitted the CORS "preflight" request, the client will be able to send actual cross-origin request.
+If the server doesn’t want to allow cross-origin access, it will respond with an error message to the client.
+
+## API
+This GEP proposes to add a new field `HTTPCORSFilter` to `HTTPRouteFilter`.
+If HTTPCORSFilter is set, then the gateway will generate the response of the preflight requests and send back it to the client directly.
+For the actual cross-origin request, the gateway will add CORS headers to the response before it is sent to the client.
+
+```golang
+const (
+    // HTTPRouteFilterCORS can be used to add CORS headers to an 
+    // HTTP response before it is sent to the client.
+    //
+    // Support in HTTPRouteRule: Extended
+    //
+    // Support in HTTPBackendRef: Extended
+    HTTPRouteFilterCORS HTTPRouteFilterType = "CORS"
+)
+
+type HTTPRouteFilter struct {
+    // CORS defines a schema for a filter that responds to the
+    // cross-origin request based on HTTP response header.
+    // 
+    // Support: Extended
+    //
+    // +optional
+    CORS *HTTPCORSFilter `json:"cors,omitempty"`
+}
+
+// HTTPCORSFilter defines a filter that responds to the
+// cross-origin request based on HTTP header.
+type HTTPCORSFilter struct {
+    // AllowOrigins indicates whether the response can be shared with 
+    // requested resource from the given `Origin`.
+    // 
+    // A wildcard indicates that the requests from all `Origin` are allowed.
```

Comment:
It's imperative that we define what the AllowOrigins field can actually contain, totally within this spec.

Here's my suggestion, that may clarify other comments about wildcards as well.

```suggestion
    // The Origin consists of a scheme and a hostname, with an optional Port, and takes the form
    // `<scheme>://<hostname>(:<port>)`.
    //
    // Valid values for scheme are: `http` and `https`.
    //
    // Valid values for port are any integer between 1 and 65535 (the list of available TCP ports).
    // Note that, if not included, port `80` is assumed for `http` scheme Origins, and port `443` is
    // assumed for `https` Origins. This may affect Origin matching.
    //
    // The hostname part of the Origin may contain the wildcard character `*`.
    // These wildcard characters behave as follows:
    //
    // * `*` is a greedy match to the _left_, including any number of DNS labels to the left of
    //    its position. This also means that `*` will include any number of period `.` characters
    //    to the left of its position.
    // * A wildcard by itself matches all hostnames.
    //
    // An Origin value that includes _only_ the `*` character indicates requests from all `Origin`s\
    // are allowed.
```

I _think_ I've got the behavior correct here, but please feel free to correct if I've made a mistake or missed something.

Reply:
Ideally, we will have CEL validation on this field to ensure that it matches the above formatting as well.

Reply:
```
Valid values for port are any integer between 1 and 65535 (the list of available TCP ports).
```
The "HTTP port" is not strictly limited to the TCP. HTTP/3 protocol uses UDP as its transport protocol.

Reply:
An origin consists of three parts: the scheme, host and port.
The HTTP "host" header is not strictly limited to the hostname.

```
host        = IP-literal / IPv4address / reg-name

IPv4address   = dec-octet "." dec-octet "." dec-octet "." dec-octet

IP-literal = "[" ( IPv6address / IPvFuture  ) "]"

 IPv6address =                            6( h16 ":" ) ls32
                  /                       "::" 5( h16 ":" ) ls32
                  / [               h16 ] "::" 4( h16 ":" ) ls32
                  / [ *1( h16 ":" ) h16 ] "::" 3( h16 ":" ) ls32
                  / [ *2( h16 ":" ) h16 ] "::" 2( h16 ":" ) ls32
                  / [ *3( h16 ":" ) h16 ] "::"    h16 ":"   ls32
                  / [ *4( h16 ":" ) h16 ] "::"              ls32
                  / [ *5( h16 ":" ) h16 ] "::"              h16
                  / [ *6( h16 ":" ) h16 ] "::"

      ls32        = ( h16 ":" h16 ) / IPv4address
                  ; least-significant 32 bits of address

      h16         = 1*4HEXDIG
                  ; 16 bits of address represented in hexadecimal

IPvFuture  = "v" 1*HEXDIG "." 1*( unreserved / sub-delims / ":" )

reg-name    = *( unreserved / pct-encoded / sub-delims )
```